### PR TITLE
New feature to search text into various websites + trigger of web browser for Pocketbook devices

### PR DESCRIPTION
**Pull request by Jeromandres**
The feature allows launching search request of selected text into a set of predefined Web Sites.
The request can be done to a default website or to a list of activated websites.

Two configuration menus allow selecting default website and activating/ deactivating websites.

As well it adds link opener for Pocketbook devices, for websites requests or direct links in documents.

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8530)

### CODE REVIEW

File: frontend/device/pocketbook/device.lua [hunk 1]
```diff
@@ -17,6 +17,29 @@ _G.POCKETBOOK_FFI = true
...
+end
+
+local function getLinkOpener()
+    if jit.os == "Linux" and isCommand(browser_app) then
+        return true, browser_app
+    elseif jit.os == "OSX" and isCommand("open") then
+        return true, "open"
+    end
+    return false
+end
```
Comment (Frenzie) [inline frontend/device/pocketbook/device.lua:41]: Quick glance only, but please remove this function.

File: frontend/dispatcher.lua [hunk 2]
```diff
@@ -90,6 +90,8 @@ local settingsList = {
     wikipedia_lookup = {category="none", event="ShowWikipediaLookup", title=_("Wikipedia lookup"), general=true},
     fulltext_search = {category="none", event="ShowFulltextSearchInput", title=_("Fulltext search"), general=true},
     file_search = {category="none", event="ShowFileSearch", title=_("File search"), general=true, separator=true},
+    websearch = { category="none", event="ShowWebLookup", title=_("Web lookup"), device=true,},
+    defaultwebsearch = { category="none", event="ShowWebDefaultLookup", title=_("Default Web lookup"), device=true,},
```
Comment (Frenzie) [inline frontend/dispatcher.lua:94]:
```suggestion
    defaultwebsearch = { category="none", event="ShowWebDefaultLookup", title=_("Default web lookup"), device=true,},
```

File: frontend/apps/reader/modules/readerwebsearch.lua [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    sub_item_table_websearch = {}
+    sub_item_table_defaultwebsearch = {}
+    
+    for websearchsite, websitedetails in pairs (websearchsites) do
+      table.insert(sub_item_table_websearch, self:genActiveWebSearchSubItem(websearchsite))
+      table.insert(sub_item_table_defaultwebsearch, self:genDefaultWebSearchSubItem(websearchsite))
+    end
+    
+    menu_items.websearch = {
+      text = _("Select Active Web Search Sites"),
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerwebsearch.lua:261]: This seems rather long, so mainly leaving this as a marker for me or others later.
```suggestion
      text = _("Select active web search sites"),
```

**Review by Frenzie — COMMENTED** [review]
Sounds interesting. Could you add a couple of screenshots if you have the time?

**Comment by poire-z** [comment]
As I get it, this:
- adds some menu to configure a list of search engine urls
- adds a button to the highlight dialog to launch a web search (with possibly added intermediate dialogs)
- launches the link opener, which is only a feature on Android, Linux, and Pocketbook; it's a dead end on Kobo and Kindle.

It extends ReaderDictionary (I understand it was probably the easiest starting point for you :) but with no real reason, as it uses nothing of its features: there is no display of results in the app itself.

All this really makes it a candidate to be just a plugin, without adding too much unrelated noise to KOReader's core (as it doesn't really add anything to the app and reading experience themselves).
Plugins can add menu items, and can add a button to the highlight dialog (none does that yet I think, but the QR code plugin did it in its early stage).
(And we'll be a little less exigeant on the coding style with a plugin - because there would be much to say :)

**Comment by Jeromandres** [comment]
Thank you for the feedback & comments.

I indeed started as a newbie from what was most similar to my intent.
I do not know about plugins yet and I will investigate.
I'll add some captures from MacOS client.

Could you share what are the next steps in general for merging? (who/ when/ how)

Regards!

**Comment by poire-z** [comment]
To move all your stuff into a plugin, I would start by just copying a simple one, say plugins/qrclipboard/, which, with the added hook to add buttons to the highlight dialog, would look like this:
```lua
local Device = require("device")
local UIManager = require("ui/uimanager")
local WidgetContainer = require("ui/widget/container/widgetcontainer")
local _ = require("gettext")

local WebSearch = WidgetContainer:new{
    name = "websearch",
    is_doc_only = true,
}

function WebSearch:init()
    self.ui.menu:registerToMainMenu(self)
    if self.ui.highlight then
        self.ui.highlight:addToHighlightDialog("23_blah", function(_self)
            return {
                text = _("Search website"),
                callback = function()
                    -- do your stuff
                end,
            }
        end)
    end
end

function WebSearch:addToMainMenu(menu_items)
    menu_items.qrclipboard = {
        text = _("QR from clipboard"),
        callback = function()
            -- add your menu stuff
        end,
    }
end

return WebSearch
```
You could nearly just migrate your own code from your ReaderWebSearch into this skeleton.
See also (for example in autostandby.koplugin/main.lua) how to make available a plugin only if specific features are available on a device, if Device:canOpenLink()).

And a few hints about our coding style vs your habits:
```lua
    for _, atitle in pairs(title_table) do
      for _, websearchsite in pairs(saved_websearchsites) do
        if (atitle == websearchsite.title) then -- no need for parenthesis with Lua
          table.insert(websearchsites, websearchsite)
        end
      end -- our preferred indentation is 4 spaces per block (not 2 like you do)
    end
                         if (ReaderWebSearch:getWebSearchFormat(websearchsite)=="UrlEncode")
                            then text = util.urlEncode(text) -- 'then' on the same line as the if
```

> Could you share what are the next steps in general for merging? (who/ when/ how)

who: us
when: when you and us are happy about the final result, so after probably a good amount of review and discussion (bear with us, we might be nitpicking, but we try to keep some standard in coding style and how things interact)
how: by you coding and correcting, and us reviewing and in the end clicking on a button :)

File: frontend/device/pocketbook/device.lua (hunk 1 above)
Comment (Jeromandres) [reply frontend/device/pocketbook/device.lua:41]: Hi

How then can I point to Pocketbook link opener script?
That is "/ebrmain/bin/webbrowser.sh"

Could you advise?
Thx

File: frontend/device/pocketbook/device.lua [hunk 4]
```diff
@@ -63,6 +86,17 @@ local PocketBook = Generic:new{
     -- Private per-model kludges
     _fb_init = function() end,
     _model_init = function() end,
+    canOpenLink = getLinkOpener,
+    openLink = function(self, link)
+        local enabled, tool = getLinkOpener()
+        logger.info("getlinkOpener: "..tostring(enabled)..", tool:"..tool)
+        logger.info("Link: "..link)
+        logger.info("Link type:"..type(link))
+        if not enabled or not tool or not link or type(link) ~= "string" then return end
```
Comment (Frenzie) [inline frontend/device/pocketbook/device.lua:95]: In reply to https://github.com/koreader/koreader/pull/8530/files#r765260999, within this structure of the code, along these lines.
```suggestion
        logger.info("Link: "..link)
        logger.info("Link type:"..type(link))
        if not isCommand(browser_app) or not link or type(link) ~= "string" then return end
```

File: frontend/device/pocketbook/device.lua [hunk 5]
```diff
@@ -17,6 +17,29 @@ _G.POCKETBOOK_FFI = true
...
+    return type(s) == "string" and s:match("*?://")
+end
+
+local function isCommand(s)
+    return os.execute("which "..s.." >/dev/null 2>&1") == 0
+end
+
+local function runCommand(command)
+      return os.execute(command) == 0
+end
```
Comment (Frenzie) [inline frontend/device/pocketbook/device.lua:32]: I suspect this is all duplicated, which means it probably wants to be refactored into util or something.

**Comment by poire-z** [comment]
Closing as not in a state to be merged.
Please re-open or create a new PR when/if you managed to make it a plugin.
Feel free to ask any questions in here if you need help though.

**Pull request closed without merging**